#-------------------------------------------------------------------------------
#pyclone-vi_005_pyclone_mutation_signatures_in_clusters.R
#Karin Isaev
#March 3rd 2021, 2021
#load R/3.5.0
#-------------------------------------------------------------------------------

#load packages and data
source("/cluster/home/kisaev/RAP_WGS/config-file.R")
library("gplots")
#library(MutationalPatterns)
library(BSgenome)
ref_genome <- "BSgenome.Hsapiens.UCSC.hg19"
library(ref_genome, character.only = TRUE)
library(NMF)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(biomaRt)
library(GenomicRanges)
library(maftools)
#Requires BSgenome object
library(BSgenome.Hsapiens.UCSC.hg19, quietly = TRUE)
library(logr)
library('pheatmap')

#data - table generated by pyclone (loci)
setwd("/cluster/projects/kridelgroup/RAP_ANALYSIS/ANALYSIS/Pyclone")

#set up patients
patients= c("LY_RAP_0001", "LY_RAP_0002", "LY_RAP_0003")

#test
#patient = "LY_RAP_0003"

z = which(read_only$Tissue_Site == "Aorta, ascending, not specified \n\n")
if(!(length(z) == 0)){
read_only$Tissue_Site[z] = "Aorta, ascending"}

#----------------------------------------------------------------------
#purpose
#----------------------------------------------------------------------

#use pyclone results clusters to identify mutation signatures associated
#with different clones

#----------------------------------------------------------------------
#data
#----------------------------------------------------------------------

#test
#patient_clonevol_results="LY_RAP_0003_all_mutations_Pyclone-VI-clonevol-results.xlsx"
#patient = "LY_RAP_0003"

#----------------------------------------------------------------------
#analysis
#----------------------------------------------------------------------

get_mut_signatures = function(patient, patient_clonevol_results){

  #read in mutation data for each cluster as obtained from pyclone
  muts <- as.data.table(read_excel(patient_clonevol_results))

  #get clusters
  clusters = unique(muts$cluster)

  #make folder to store plots for patient
  mainDir <- "/cluster/projects/kridelgroup/RAP_ANALYSIS/ANALYSIS/Pyclone/clusters_mutation_signatures"
  subDir <- patient

  dir.create(file.path(mainDir, subDir), showWarnings = FALSE)
  setwd(file.path(mainDir, subDir))

  #for each cluster gather mutations and their details from main mutation
  #file
  get_cluster_muts = function(clust){

    cluster_muts = filter(muts, cluster==clust)
    mut.merged = as.data.table(filter(read_only, STUDY_PATIENT_ID == patient, mut_id %in% cluster_muts$mut_id))
    mut.merged = unique(mut.merged[,c("mut_id", "ensgene", "POS", "CHROM", "symbol", "REF","ALT",
    "Gene.ensGene", "GeneDetail.ensGene", "ExonicFunc.ensGene", "Func.ensGene")])
    mut.merged$Variant_Classification = paste(mut.merged$Func.ensGene, mut.merged$ExonicFunc.ensGene)

    #maftools
    #Mandatory fields: Hugo_Symbol, Chromosome, Start_Position, End_Position,
    #Reference_Allele, Tumor_Seq_Allele2,
    #Variant_Classification, Variant_Type and Tumor_Sample_Barcode.
    colnames(mut.merged)[3] = "Start_Position"
    mut.merged$End_Position = mut.merged$Start_Position
    colnames(mut.merged)[4] = "Chromosome"
    colnames(mut.merged)[5] = "Hugo_Symbol"
    colnames(mut.merged)[6:7] = c("Reference_Allele", "Tumor_Seq_Allele2")
    mut.merged$Tumor_Sample_Barcode = patient

    #prepare input for maftools
    mut.merged$Variant_Classification[mut.merged$Variant_Classification == "exonic frameshift_deletion"] = "Frame_Shift_Del"
    mut.merged$Variant_Classification[mut.merged$Variant_Classification == "exonic frameshift_insertion"] = "Frame_Shift_Ins"
    mut.merged$Variant_Classification[mut.merged$Variant_Classification == "exonic nonsynonymous_SNV"] = "Missense_Mutation"
    mut.merged$Variant_Classification[mut.merged$Variant_Classification == "exonic nonframeshift_deletion"] = "In_Frame_Del"
    mut.merged$Variant_Classification[mut.merged$Variant_Classification == "exonic stopgain"] = "Nonsense_Mutation"
    mut.merged$Variant_Classification[mut.merged$Variant_Classification == "exonic nonframeshift_insertion"] = "In_Frame_Ins"
    mut.merged$Variant_Classification[mut.merged$Variant_Classification == "exonic nonframeshift_substitution"] = "Inframe_INDEL"
    mut.merged$Variant_Classification[mut.merged$Variant_Classification == "exonic stoploss"] = "Nonstop_Mutation"

    #make rest of mutations "misssense" so can use full dataset for signature analysis
    mut.merged$Variant_Classification[!(mut.merged$Variant_Classification %in% c("Frame_Shift_Del", "Frame_Shift_Ins",
        "Missense_Mutation", "In_Frame_Del", "Nonsense_Mutation", "In_Frame_Ins", "Inframe_INDEL", "Nonstop_Mutation"))] = "Missense_Mutation"

    mut.merged[,ref_alt_diff := nchar(Reference_Allele) - nchar(Tumor_Seq_Allele2)]
    mut.merged[, Variant_Type := ifelse(ref_alt_diff == 0 , yes = "SNP", no = ifelse(ref_alt_diff < 0 , yes = "INS", no = "DEL"))]
    #save as maf object
    write.table(mut.merged, file="tem_maf_file.txt", quote=F, row.names=F, sep="\t")
    muts_maf = read.maf(maf = "tem_maf_file.txt")

    pdf(paste(clust, "mutation_signature_analysis_plots.pdf", sep="_"))

    laml.titv = titv(maf = muts_maf, plot = FALSE, useSyn = TRUE)

    #plot titv summary
    plotTiTv(res = laml.titv)

    #rainfall plot
    rainfallPlot(maf = muts_maf, detectChangePoints = TRUE, pointSize = 0.4)

    #mutation signatures analysis
    laml.tnm = trinucleotideMatrix(maf = muts_maf, ref_genome = "BSgenome.Hsapiens.UCSC.hg19")

    #library('NMF')
    #laml.sign = estimateSignatures(mat = laml.tnm, nMin = 2, nTry = 3, nrun = 2, pConstant = 0.01)

    laml.sig = extractSignatures(mat = laml.tnm, n = 2, pConstant = 1e-9)
    laml.v3.cosm = compareSignatures(nmfRes = laml.sig, sig_db = "SBS")
    pheatmap::pheatmap(mat = laml.v3.cosm$cosine_similarities, cluster_rows = FALSE, main = "cosine similarity against validated signatures")
    maftools::plotSignatures(nmfRes = laml.sig, title_size = 1.2, sig_db = "SBS")
    dev.off()

    #save result
    sigs=as.data.table(laml.v3.cosm$best_match)
    sigs$patient = patient
    sigs$clust=clust
    return(sigs)
  }

  all_clusts = as.data.table(ldply(llply(clusters, get_cluster_muts)))
  setwd("/cluster/projects/kridelgroup/RAP_ANALYSIS/ANALYSIS/Pyclone")
  return(all_clusts)
}

p001 = get_mut_signatures("LY_RAP_0001", "LY_RAP_0001_all_mutations_Pyclone-VI-clonevol-results.xlsx")
p002 = get_mut_signatures("LY_RAP_0002", "LY_RAP_0002_all_mutations_Pyclone-VI-clonevol-results.xlsx")
p003 = get_mut_signatures("LY_RAP_0003", "LY_RAP_0003_all_mutations_Pyclone-VI-clonevol-results.xlsx")

#combine all results
all_results = rbind(p001, p002, p003)
saveRDS(all_results, file="all_pyclone_clusters_signatures_results.rds")

#-----
#DONE-
#-----
