#!/bin/bash
#
#SBATCH -N 1 # Ensure that all cores are on one machine
#SBATCH -p himem
#SBATCH --mem=61440M
#SBATCH -t 5-00:00 # Runtime in D-HH:MM
#SBATCH -J strelka
#SBATCH -c 8
#SBATCH --array=0-26 # job array index

#need to run Mutect2 on all bam files from tumour samples
#author: Karin Isaev
#date started: June 25, 2019
#date updated: October 30th, 2020

module load java/8  #8
module load samtools
module load python3
module load gatk
module load annovar
module load vt

#pwd
cd /cluster/projects/kridelgroup/RAP_ANALYSIS/STRELKA_WORKDIR

#collect all vcf files generated by strelka for each sample
#ls */*/*/somatic.indels.vcf.gz > strelka_default_indels_output

names=($(cat strelka_default_indels_output))
echo ${names[${SLURM_ARRAY_TASK_ID}]}

#gatk IndexFeatureFile \
#     -F ${names[${SLURM_ARRAY_TASK_ID}]}

MYVAR=${names[${SLURM_ARRAY_TASK_ID}]}
tum_name=${MYVAR%/results/variants/somatic.indels.vcf.gz*}
patient_name=${tum_name##*STRELKA_WORKDIR_}

fasta=/cluster/projects/kridelgroup/RAP_ANALYSIS/human_g1k_v37_decoy.fasta #from gatk resource bundle

gatk SelectVariants \
   -R $fasta \
   -V ${names[${SLURM_ARRAY_TASK_ID}]} \
   -O strelka_filtered/${patient_name}_strelka_selectvariants.indels.vcf.gz \
   --exclude-filtered true \
   --select-type-to-exclude SNP \
   --exclude-intervals Y \

#normalize file
input_vcf=strelka_filtered/${patient_name}_strelka_selectvariants.indels.vcf.gz

#normalize variants, send to standard out and remove duplicates.
vt normalize ${input_vcf} -r $fasta -o ${input_vcf}.normalized.vcf.gz

#DONE
